http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -416,4 +484,112 @@ private static boolean hasScope(String scopeParam, String targetScope) {
+            jwtClaims.setClaim(parameter.getName(), parameter.getValue());
+        }
+
+        // sign JWT first before encrypting
+        JsonWebSignature signedRequest = signRequest(jwtClaims);
+
+        // Encrypting optional
+        if (deployment.getRequestEncryptAlgorithm() != null && !deployment.getRequestEncryptAlgorithm().isEmpty() &&
+            deployment.getRequestEncryptEncValue() != null && !deployment.getRequestEncryptEncValue().isEmpty()) {
+            return encryptRequest(signedRequest).getCompactSerialization();
+        } else {
+            return signedRequest.getCompactSerialization();
+        }
+    }
+
+    public KeyPair getkeyPair() throws IOException {
+        if (deployment.getClientKeyStoreFile().contains(PROTOCOL_CLASSPATH)) {
+            deployment.setClientKeyStoreFile(deployment.getClientKeyStoreFile().replace(PROTOCOL_CLASSPATH, Objects.requireNonNull(Thread.currentThread().getContextClassLoader().getResource("")).getPath()));
+        }
+        if (!deployment.getRequestSignatureAlgorithm().equals(NONE) && deployment.getClientKeyStoreFile() == null){
+            throw log.invalidKeyStoreConfiguration();
+        } else {
+            return loadKeyPairFromKeyStore(deployment.getClientKeyStoreFile(),
+                    deployment.getClientKeyStorePassword(), deployment.getClientKeyPassword(),
+                    deployment.getClientKeyAlias(), deployment.getClientKeystoreType());
+        }
+    }
+
+    public JsonWebSignature signRequest(JwtClaims jwtClaims) throws Exception {
+        JsonWebSignature jsonWebSignature = new JsonWebSignature();
+        jsonWebSignature.setPayload(jwtClaims.toJson());
+
+        if (deployment.getRequestSignatureAlgorithm().contains(NONE)) { //unsigned
+            jsonWebSignature.setAlgorithmConstraints(AlgorithmConstraints.NO_CONSTRAINTS);
+            jsonWebSignature.setAlgorithmHeaderValue(NONE);
+        } else if (deployment.getRequestSignatureAlgorithm().contains(HS_256)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_384)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_512)) { //signed with symmetric key
+            jsonWebSignature.setAlgorithmHeaderValue(deployment.getRequestSignatureAlgorithm());
+            Key key = new HmacKey(deployment.getResourceCredentials().get("secret").toString().getBytes(StandardCharsets.UTF_8));   //the client secret is a shared secret between the server and the client
+            jsonWebSignature.setDoKeyValidation(false);     //skips validation so that size of the secret does not matter
+            jsonWebSignature.setKey(key);
+        } else { //signed with asymmetric key
+            KeyPair keyPair = getkeyPair();
+            jsonWebSignature.setKey(keyPair.getPrivate());
+            jsonWebSignature.setAlgorithmConstraints(new AlgorithmConstraints(PERMIT, deployment.getRequestSignatureAlgorithm()));
```

Comment:
Just to check, why do we need to specify this?

Reply:
I think I added it because I thought just like NONE signing algorithm, there would be encryption algorithms that JWE would complain about, but that's not the case. So I will remove it. 
Thanks for catching it :)